File: Lists/StudentSupportOffice.cs
namespace Lists;

public class StudentSupportOffice
{
    public IReadOnlyDictionary<string, Module> CreateDefaultModules(IComparer<Student>? comparer = null)
    {
        // Creates the required modules with the specified comparer.
        // Using a dictionary to store modules as it makes it easier and faster to reference by ModuleID.
        var modules = new Dictionary<string, Module>
        {
            ["CIS2344"] = new("CIS2344", comparer),
            ["CIS2360"] = new("CIS2360", comparer),
            ["CIM2130"] = new("CIM2130", comparer)
        };

        return modules;
    }

    public IReadOnlyDictionary<string, Student> CreateDefaultStudents()
    {
        const int studentCount = 5;

        // Using a dictionary to store students as it makes it easier and faster to reference by StudentID.
        var students = new Dictionary<string, Student>(studentCount);

        // Simple Student Creation

        for (var i = 1; i < studentCount + 1; i++)
        {
            var studentId = $"U{i:0000000}";
            students[studentId] = new Student(studentId);
        }

        return students;
    }
    
    public void DefaultEnrollment(IReadOnlyDictionary<string, Module> modules, IReadOnlyDictionary<string, Student> students)
    {
        // The default enrollment criteria as specified in the assignment.
        modules["CIS2344"].Enroll(students["U0000001"]);
        modules["CIS2344"].Enroll(students["U0000004"]);
        
        modules["CIS2360"].Enroll(students["U0000001"]);
        modules["CIS2360"].Enroll(students["U0000003"]);
        modules["CIS2360"].Enroll(students["U0000005"]);
        
        modules["CIM2130"].Enroll(students["U0000003"]);
        modules["CIM2130"].Enroll(students["U0000004"]);
        modules["CIM2130"].Enroll(students["U0000005"]);
    }

    public void EnrollmentChanges(IReadOnlyDictionary<string, Module> modules,
        IReadOnlyDictionary<string, Student> students)
    {
        // The end of enrollment criteria as specified in the assignment.

        modules["CIS2344"].Enroll(students["U0000002"]);
        modules["CIS2344"].Enroll(students["U0000003"]);
        // Evidence that duplicate students can't exist.
        modules["CIS2344"].Enroll(students["U0000004"]);
        modules["CIS2344"].Enroll(students["U0000005"]);
        
        modules["CIS2360"].Enroll(students["U0000002"]);
        modules["CIS2360"].Disenroll(students["U0000003"]);

        modules["CIM2130"].Enroll(students["U0000001"]);
        modules["CIM2130"].Disenroll(students["U0000005"]);
    }

    public IReadOnlyList<Grade> CreateDefaultGrades(IReadOnlyDictionary<string, Module> modules,
        IReadOnlyDictionary<string, Student> students)
    {
        return new List<Grade>(8)
        {
            new(modules["CIS2344"], students["U0000001"], 70),
            new(modules["CIS2344"], students["U0000004"], 66),

            new(modules["CIS2360"], students["U0000001"], 74),
            new(modules["CIS2360"], students["U0000003"], 38),
            new(modules["CIS2360"], students["U0000005"], 52),

            new(modules["CIM2130"], students["U0000003"], 91),
            new(modules["CIM2130"], students["U0000004"], 38),
            new(modules["CIM2130"], students["U0000005"], 29)
        };
    }

    public void RegisterGrades(IEnumerable<Grade> grades)
    {
        foreach (var grade in grades)
            grade.Student.AddNewGrade(grade);
    }

    public void UpdateGradeScore(IReadOnlyList<Grade> grades)
    {
        grades[0].Score = 71;
        grades[2].Score = 55;
        grades[4].Score = 21;
    }

    public void AmendGradeViaAddGrade(IReadOnlyList<Grade> grades)
    {
        grades[1].Student.AddNewGrade(new Grade(grades[1].Module, grades[1].Student, 43));
    }

    public void RemoveGrades(IReadOnlyList<Grade> grades)
    {
        grades[1].Destroy();
        grades[3].Destroy();
        grades[5].Destroy();
    }
}

File: CIS2206/src/Practical_11/StudentSupportOffice.java
package Practical_11;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;

public class StudentSupportOffice {
    public Map<String, Module> getAndCreateDefaultModules(Comparator<Student> comparator) {
        // Creates the required modules with the specified comparator.
        // Using a hash map to store modules as it makes it easier and faster to reference by ModuleID.
        var modules = new HashMap<String, Module>(3);
        modules.put("CIS2206", new Module("CIS2206", comparator));
        modules.put("CIS2360", new Module("CIS2360", comparator));
        modules.put("CIS2205", new Module("CIS2205", comparator));

        return modules;
    }

    public Map<String, Student> getAndCreateDefaultStudents() {
        // Using a hash map to store students as it makes it easier and faster to reference by StudentID.
        var students = new HashMap<String, Student>(5);

        // Complex Student Creation
        // Creates Students with a unique ID, in addition to basic attributes.
        var student = new Student("U0000001", "Hall", "Liam", 22);
        students.put(student.getStudentId(), student);

        student = new Student("U0000002", "Sinla", "Nari", 30);
        students.put(student.getStudentId(), student);

        student = new Student("U0000003", "Conhee", "Drea", 20);
        students.put(student.getStudentId(), student);

        student = new Student("U0000004", "Foster", "Harry", 21);
        students.put(student.getStudentId(), student);

        student = new Student("U0000005", "Musk", "Elon", 51);
        students.put(student.getStudentId(), student);

        return students;
    }

    public void defaultEnrollment(Map<String, Module> modules, Map<String, Student> students)
    {
        // The default enrollment criteria as specified in the assignment.
        modules.get("CIS2206").enroll(students.get("U0000001"));
        modules.get("CIS2206").enroll(students.get("U0000005"));

        modules.get("CIS2360").enroll(students.get("U0000001"));
        modules.get("CIS2360").enroll(students.get("U0000003"));
        modules.get("CIS2360").enroll(students.get("U0000004"));

        modules.get("CIS2205").enroll(students.get("U0000002"));
        modules.get("CIS2205").enroll(students.get("U0000004"));
        modules.get("CIS2205").enroll(students.get("U0000005"));
    }

    public void enrollmentChanges(Map<String, Module> modules, Map<String, Student> students)
    {
        // The end of enrollment criteria as specified in the assignment.
        modules.get("CIS2206").enroll(students.get("U0000003"));
        modules.get("CIS2206").enroll(students.get("U0000004"));
        modules.get("CIS2206").disenroll(students.get("U0000005"));

        // Evidence that duplicate students can't exist.
        modules.get("CIS2206").enroll(students.get("U0000003"));

        modules.get("CIS2360").enroll(students.get("U0000002"));
        modules.get("CIS2360").enroll(students.get("U0000005"));

        modules.get("CIS2205").enroll(students.get("U0000001"));
        modules.get("CIS2205").enroll(students.get("U0000003"));
        modules.get("CIS2205").disenroll(students.get("U0000002"));
        modules.get("CIS2205").disenroll(students.get("U0000005"));
    }
}
